Build failure: pkgsLLVM.llvmPackages_19.compiler-rt

### Steps To Reproduce

Steps to reproduce the behavior:
1. Use
```nix
	pkgs = (import nixpkgs {
		system = builtins.currentSystem;

		overlays = [
			(final: super: {
				llvmPackages = super.llvmPackages_git;
			})
		];
	}).pkgsLLVM;
```
(or try building pkgsLLVM.llvmPackages_git.compiler-rt)

### Build log

```
compiler-rt> CMake Error at cmake/Modules/AddCompilerRT.cmake:357 (add_library):
compiler-rt>   Error evaluating generator expression:
compiler-rt>
compiler-rt>     $<TARGET_OBJECTS:RTSanitizerCommon.x86_64>
compiler-rt>
compiler-rt>   Objects of target "RTSanitizerCommon.x86_64" referenced but no such target
compiler-rt>   exists.
compiler-rt> Call Stack (most recent call first):
compiler-rt>   lib/ctx_profile/CMakeLists.txt:25 (add_compiler_rt_runtime)
compiler-rt>
compiler-rt>
compiler-rt> CMake Error at cmake/Modules/AddCompilerRT.cmake:357 (add_library):
compiler-rt>   No SOURCES given to target: clang_rt.ctx_profile-x86_64
compiler-rt> Call Stack (most recent call first):
compiler-rt>   lib/ctx_profile/CMakeLists.txt:25 (add_compiler_rt_runtime)
compiler-rt>
compiler-rt>
compiler-rt> CMake Generate step failed.  Build files cannot be regenerated correctly.
....
```

### Additional context

It seems like useLLVM causes the sanitizers to be disabled, not sure why this is the case as LLVM can rebuild itself with sanitizers enabled (outside of nixpkgs).

See https://github.com/NixOS/nixpkgs/blob/09da1488a3b3bd46ef1150946539b3ee6e738af6/pkgs/development/compilers/llvm/common/compiler-rt/default.nix#L87C1-L88C41

### Notify maintainers

### Metadata

> Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)
```

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

Ok, removing those useLLVM checks resulted in this:
```
compiler-rt> /build/compiler-rt-src-19.0.0-unstable-2024-07-21/compiler-rt/lib/sanitizer_common/sanitizer_errno.cpp:19:10: fatal error: 'errno.h' file not found
compiler-rt>    19 | #include <errno.h>
compiler-rt>       |          ^~~~~~~~~

```

> Unable to reproduce this.

It's reproducible by building `llvmPackages_git.compiler-rt-no-libc` at version `20.0.0-unstable-2024-07-28`:

```
nix-repl> :b inputs.nixpkgs.legacyPackages.x86_64-linux.llvmPackages_git.compiler-rt-no-libc
error: builder for '/nix/store/xwg6nlz8m0bha4p1aklapczzy2lpab9r-compiler-rt-20.0.0-unstable-2024-07-28.drv' failed with exit code 1;
       last 25 log lines:
       >
       >   Objects of target "RTSanitizerCommon.x86_64" referenced but no such target
       >   exists.
       > Call Stack (most recent call first):
       >   lib/ctx_profile/CMakeLists.txt:25 (add_compiler_rt_runtime)
       >
       > 
       > CMake Error at cmake/Modules/AddCompilerRT.cmake:357 (add_library):
       >   Error evaluating generator expression:
       >
       >     $<TARGET_OBJECTS:RTSanitizerCommon.x86_64>
       >
       >   Objects of target "RTSanitizerCommon.x86_64" referenced but no such target
       >   exists.
       > Call Stack (most recent call first):
       >   lib/ctx_profile/CMakeLists.txt:25 (add_compiler_rt_runtime)
       >
       > 
       > CMake Error at cmake/Modules/AddCompilerRT.cmake:357 (add_library):
       >   No SOURCES given to target: clang_rt.ctx_profile-x86_64
       > Call Stack (most recent call first):
       >   lib/ctx_profile/CMakeLists.txt:25 (add_compiler_rt_runtime)
       >
       > 
       > CMake Generate step failed.  Build files cannot be regenerated correctly.
       For full logs, run 'nix log /nix/store/xwg6nlz8m0bha4p1aklapczzy2lpab9r-compiler-rt-20.0.0-unstable-2024-07-28.drv'.
```

Building version `19.0.0-unstable-2024-07-21` works fine:
```
nix-repl> :b llvmPackages_git.compiler-rt-no-libc

This derivation produced the following outputs:
  dev -> /nix/store/s34myijw3y2fkzkzc8zic9hkyxp8s3fi-compiler-rt-libc-19.0.0-unstable-2024-07-21-dev
  out -> /nix/store/727m6g4znkgds0b7bifwh7kcawsdhrp1-compiler-rt-libc-19.0.0-unstable-2024-07-21
```

--- a/pkgs/development/compilers/llvm/common/compiler-rt/default.nix
+++ b/pkgs/development/compilers/llvm/common/compiler-rt/default.nix
@@ -39,6 +39,7 @@ let
   haveLibcxx = stdenv.cc.libcxx != null;
   isDarwinStatic = stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isStatic && lib.versionAtLeast release_version "16";
   inherit (stdenv.hostPlatform) isMusl isAarch64;
+  noSanitizers = !haveLibc || bareMetal || isMusl || isDarwinStatic;
 
   baseName = "compiler-rt";
   pname = baseName + lib.optionalString (haveLibc) "-libc";
@@ -94,7 +95,7 @@ stdenv.mkDerivation ({
     "-DCOMPILER_RT_BUILD_LIBFUZZER=OFF"
   ] ++ lib.optionals (useLLVM && haveLibc) [
     "-DCOMPILER_RT_BUILD_SANITIZERS=ON"
-  ] ++ lib.optionals (!haveLibc || bareMetal || isMusl || isDarwinStatic) [
+  ] ++ lib.optionals (noSanitizers) [
     "-DCOMPILER_RT_BUILD_SANITIZERS=OFF"
   ] ++ lib.optionals ((useLLVM && !haveLibcxx) || !haveLibc || bareMetal || isMusl || isDarwinStatic) [
     "-DCOMPILER_RT_BUILD_XRAY=OFF"
@@ -131,6 +132,8 @@ stdenv.mkDerivation ({
     "-DCOMPILER_RT_ENABLE_IOS=OFF"
   ]) ++ lib.optionals (lib.versionAtLeast version "19" && stdenv.isDarwin && lib.versionOlder stdenv.hostPlatform.darwinMinVersion "10.13") [
     "-DSANITIZER_MIN_OSX_VERSION=10.10"
+  ]  ++ lib.optionals (noSanitizers && lib.versionAtLeast release_version "19") [
+    "-DCOMPILER_RT_BUILD_CTX_PROFILE=OFF"
   ];
 
   outputs = [ "out" "dev" ];

